pkg/validation.go

```diff
@@ -0,0 +1,251 @@
+			result.Valid = false
+			result.Errors = append(result.Errors, ValidationError{
+				Message: fmt.Sprintf("HCL parsing error: %s", diags.Error()),
+				File:    tfFile,
+			})
+			continue
+		}
+		allFiles = append(allFiles, fileWithPath{file: file, path: tfFile})
+	}
+
+	// Validate requirements across all files
+	validationErrors := validateRequirements(allFiles, config)
+	result.Errors = append(result.Errors, validationErrors...)
+
+	if len(validationErrors) > 0 {
+		result.Valid = false
+	}
+
+	return result
+}
+
+// findTerraformFiles finds all .tf files in the given directory
+func findTerraformFiles(dir string) ([]string, error) {
+	var tfFiles []string
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() && strings.HasSuffix(strings.ToLower(info.Name()), ".tf") {
+			tfFiles = append(tfFiles, path)
+		}
+
+		return nil
+	})
+
+	return tfFiles, err
+}
+
+// validateRequirements checks that all required elements are present
+func validateRequirements(files []fileWithPath, config ValidationConfig) []ValidationError {
+	var errors []ValidationError
+
+	// Track what we've found
+	foundVariables := make(map[string]bool)
+	foundProviders := make(map[string]bool)
+	foundS3Backend := false
+	foundTerraformBlock := false
+
+	// Parse each file
+	for _, fileInfo := range files {
+		content, _, diags := fileInfo.file.Body.PartialContent(&hcl.BodySchema{
+			Blocks: []hcl.BlockHeaderSchema{
+				{Type: "terraform"},
+				{Type: "variable", LabelNames: []string{"name"}},
+				{Type: "provider", LabelNames: []string{"name"}},
+			},
+		})
+
+		if diags.HasErrors() {
+			errors = append(errors, ValidationError{
+				Message: fmt.Sprintf("Error parsing HCL structure: %s", diags.Error()),
+				File:    fileInfo.path,
+			})
+			continue
+		}
+
+		// Check terraform blocks
+		for _, block := range content.Blocks {
+			switch block.Type {
+			case "terraform":
+				foundTerraformBlock = true
+				terraformContent, _, _ := block.Body.PartialContent(&hcl.BodySchema{
+					Blocks: []hcl.BlockHeaderSchema{
+						{Type: "backend", LabelNames: []string{"type"}},
+						{Type: "required_providers"},
+					},
+				})
+
+				// Check for S3 backend
+				for _, backendBlock := range terraformContent.Blocks {
+					if backendBlock.Type == "backend" && len(backendBlock.Labels) > 0 && backendBlock.Labels[0] == "s3" {
+						foundS3Backend = true
+					}
+				}
+
+				// Check required_providers
+				for _, reqProvidersBlock := range terraformContent.Blocks {
+					if reqProvidersBlock.Type == "required_providers" {
+						attrs, _ := reqProvidersBlock.Body.JustAttributes()
+						for _, provider := range config.RequiredProviders {
+							if _, exists := attrs[provider.LocalName]; exists {
+								foundProviders[provider.LocalName] = true
+							}
+						}
+					}
+				}
+
+			case "variable":
+				if len(block.Labels) > 0 {
+					foundVariables[block.Labels[0]] = true
```

Comment:
Somewhat confused by the interaction between block labels and found resources. 

To confirm, because for both `providers` and `variables` the original `Blocks` contain `LabelNames` slices with a single `name` value, it is safe to assume indexing `0` of `block.Labels` is the correct value here?

Reply:
Upon review, in a `provider` example of `backend "s3" {}` or variable example of `variable "foo" {}`, `Labels` will indeed contain a single value, which is `s3` or `foo`. LGTM